lib/Sema/CSDiagnostics.cpp

```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
   // we find a node in the lvalue path that is problematic, this returns it.
   auto immInfo = resolveImmutableBase(destExpr);
 
-  // Otherwise, we cannot resolve this because the available setter candidates
-  // are all mutating and the base must be mutating.  If we dug out a
-  // problematic decl, we can produce a nice tailored diagnostic.
-  if (auto *VD = dyn_cast_or_null<VarDecl>(immInfo.second)) {
-    std::string message = "'";
-    message += VD->getName().str().str();
-    message += "'";
-
-    auto type = getType(immInfo.first);
-    auto bgt = type ? type->getAs<BoundGenericType>() : nullptr;
-
-    if (bgt && bgt->getDecl() == getASTContext().getKeyPathDecl())
-      message += " is a read-only key path";
-    else if (VD->isCaptureList())
-      message += " is an immutable capture";
-    else if (VD->isImplicit())
-      message += " is immutable";
-    else if (VD->isLet())
-      message += " is a 'let' constant";
-    else if (!VD->isSettable(DC))
-      message += " is a get-only property";
-    else if (!VD->isSetterAccessibleFrom(DC))
-      message += " setter is inaccessible";
-    else {
-      message += " is immutable";
-    }
-
-    emitDiagnostic(Loc, DeclDiagnostic, message)
-        .highlight(immInfo.first->getSourceRange());
-
-    // If there is a masked instance variable of the same type, emit a
-    // note to fixit prepend a 'self.'.
-    if (auto typeContext = DC->getInnermostTypeContext()) {
-      UnqualifiedLookup lookup(VD->getFullName(), typeContext,
-                               getASTContext().getLazyResolver());
-      for (auto &result : lookup.Results) {
-        const VarDecl *typeVar = dyn_cast<VarDecl>(result.getValueDecl());
-        if (typeVar && typeVar != VD && typeVar->isSettable(DC) &&
-            typeVar->isSetterAccessibleFrom(DC) &&
-            typeVar->getType()->isEqual(VD->getType())) {
-          // But not in its own accessor.
-          auto AD =
-              dyn_cast_or_null<AccessorDecl>(DC->getInnermostMethodContext());
-          if (!AD || AD->getStorage() != typeVar) {
-            emitDiagnostic(Loc, diag::masked_instance_variable,
-                           typeContext->getSelfTypeInContext())
-                .fixItInsert(Loc, "self.");
+  Optional<OverloadChoice> choice = immInfo.second;
+
+  // Attempt diagnostics based on the overload choice.
+  if (choice.hasValue()) {
+
+    if (!choice->isDecl()) {
+      if (choice->getKind() == OverloadChoiceKind::KeyPathApplication &&
+          !isa<ApplyExpr>(immInfo.first)) {
+        std::string message = "the key path";
+        if (auto *SE = dyn_cast<SubscriptExpr>(immInfo.first)) {
+          if (auto *tupleExpr = dyn_cast<TupleExpr>(SE->getIndex())) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(tupleExpr->getElement(0))) {
```

Comment:
These nested ifs for all these `dyn_cast`s are kinda ugly, but I'm not sure if there's a nicer way to write this. Any suggestions would be appreciated 🙂

Reply:
You can combine some of them together. For example:

```c++
if (choice.hasValue() && !choice->isDecl()) { ... }
```

You can also write the casts as:

```c++
auto *SE = dyn_cast<SubscriptExpr>(immInfo.first);
if (!SE) return;

auto *TE = dyn_cast<TupleExpr>(SE->getIndex());
if (!TE) return;

...
```

Reply:
I don't think i can combine `choice.hasValue()` with `!choice->isDecl()`, cause the following if depends on it being a decl and is also inside the `if (choice.hasValue()) {...}`. Also, in this case, we sadly can't just return if the cast fails because we want to keep going but just without setting the message. I guess there's not a super clean way to write it in this case so I'll keep it as is for now.

Reply:
I was referring to line 984

Reply:
Oh, you're right. You can skip the `return`'s and use `dyn_cast_or_null` instead. If any of the casts fails, your final expression/decl will also be `null`.

```c++
auto *SE = dyn_cast_or_null<SubscriptExpr>(immInfo.first);
auto *TE = dyn_cast_or_null<TupleExpr>(SE ? SE->getIndex() : nullptr);
...
```

Not a big fan of this though so I guess it's better to stick with the ifs!

Reply:
Yeah, I suppose so. I hate seeing that many levels of nesting, but guess there's not anything super nice we can do there.

Reply:
Another way of doing this would be to extract this into `Expr *getKeyPathArgument(SubscriptExpr *expr)` either static function or a closure here.